progressive_rollout_demo/lib/main.dart

```diff
@@ -0,0 +1,144 @@
+import 'package:flutter/material.dart';
+import 'package:package_info_plus/package_info_plus.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+import 'package:shorebird_code_push/shorebird_code_push.dart';
+
+const groupKey = 'groupNumber';
+const firestoreCollectionName = 'rollouts';
+const firebaseDocumentName = 'rollouts';
+
+/// Assigns a random group number (1-100) to the user if one has not already
+/// been assigned.
+Future<void> assignGroupNumberIfNeeded() async {
+  final prefs = await SharedPreferences.getInstance();
+  if (!prefs.containsKey(groupKey)) {
+    final groupNumber = Random().nextInt(100) + 1;
+    await prefs.setInt(groupKey, groupNumber);
+  }
+}
+
+Future<void> main() async {
+  WidgetsFlutterBinding.ensureInitialized();
+
+  await Firebase.initializeApp();
+  await assignGroupNumberIfNeeded();
+
+  runApp(const MyApp());
+}
+
+class MyApp extends StatelessWidget {
+  const MyApp({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      theme: ThemeData(
+        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
+        useMaterial3: true,
+      ),
+      home: const MyHomePage(),
+    );
+  }
+}
+
+class MyHomePage extends StatefulWidget {
+  const MyHomePage({super.key});
+
+  @override
+  State<MyHomePage> createState() => _MyHomePageState();
+}
+
+class _MyHomePageState extends State<MyHomePage> {
+  final _updater = ShorebirdUpdater();
+
+  int? _groupNumber;
+
+  /// Returns the current release version (e.g., 1.0.0+1).
+  Future<String> _releaseVersion() async {
+    final packageInfo = await PackageInfo.fromPlatform();
+    return '${packageInfo.version}+${packageInfo.buildNumber}';
+  }
+
+  /// Fetches the rollout percentage for the current release version
+  /// (e.g., 1.0.0+1) from Firestore.
+  ///
+  /// Shorebird employees can view the Firestore data in the console at:
+  /// https://console.firebase.google.com/u/0/project/progressive-rollout-demo/firestore/databases/-default-/data/~2Frollouts~2Frollouts
+  Future<int?> _fetchRolloutPercentage() async {
+    final collection = await FirebaseFirestore.instance
+        .collection(firestoreCollectionName)
+        .get();
+    final releaseVersion = await _releaseVersion();
+    return collection.docs.first.data()[releaseVersion] as int?;
+  }
+
+  /// Determines the update track (beta or stable) based on the group number
+  /// and rollout percentage. If a user's group number is less than or equal to
+  /// the rollout percentage, they are assigned to the beta track. Otherwise,
+  /// they are assigned to the stable track. If there is no rollout percentage
+  /// for the current release version, all users are on the stable track.
+  Future<UpdateTrack> _updateTrack() async {
+    // If we haven't set a group number yet, assume we're in the stable track.
+    // This should never happen.
+    if (_groupNumber == null) return UpdateTrack.stable;
```

Comment:
Couldn't we simplify this further by having a global late final groupNumber that gets optionally assigned and initialized so that this code can be deleted?

Reply:
We could. I'm trying to keep this contained to the widget.

Reply:
If we are going to keep it like this (nullable) then I'd probably throw an Exception if groupNumber is null instead of defaulting to the stable track.

Reply:
I think we're bikeshedding at this point

Reply:
I just think as a sample which we should assume lots of people directly copy it’d be nice not to have the potential for subtle bugs to creep in so easily. If someone replaced SharedPreferences with something else that took much longer to query this would very easily manifest as non-deterministic behavior.

Reply:
If they copy the code and then change it, there's not really anything we can do there.

There are probably a million things like this that we could change. I don't think we're gaining a lot of value by making this code bulletproof.

Reply:
We shouldn't have sample code that contains race conditions and could lead to non-deterministic behavior imo. This is a real bug in the implementation because the behavior of the code depends on the timing of the `groupNumber` read operation. I'm happy to make the change if you prefer -- I don't think it would be a big refactor fwiw.

Reply:
You're welcome to make the change if you'd like. I think we've already sunk way too much time into sample code that could be spent on other, more valuable pursuits.